### ISSUE TITLE
Getting NAN loss while training with 1 class

### ISSUE THREAD

**Issue by Robin2091** [zzh8829/yolov3-tf2#86]
I am training with 1 class and during training, I receive NAN for the loss value. I am training on a CPU with 15 images as a test( I will be using a cloud GPU for the actual training). I changed the loss function to binary cross-entropy as the comment says on the models.py file. Here are the settings I have: 
![image](https://user-images.githubusercontent.com/57192713/67885993-d5affe80-fb05-11e9-89e4-7d5807758d10.png)
Here is what I get when I run "python train.py":
![image](https://user-images.githubusercontent.com/57192713/67886224-5d960880-fb06-11e9-8415-adb815f28764.png)
This is some of the training data I am using for the 1 class. I created a script that reads the tfrecords and displays the image with the bounding boxes on top . 
![image](https://user-images.githubusercontent.com/57192713/67891097-1eb88080-fb0f-11e9-9c26-c4f89ee7efda.png)
![image](https://user-images.githubusercontent.com/57192713/67891128-2c6e0600-fb0f-11e9-96f7-c03f8d3b04e1.png)
So for some reason, it doesn't give me NAN loss values for the first batch of epoch 1. 
Although, when I train on another custom dataset with 15 examples which has 3 classes, training works fine and I don't receive any NAN loss values. 
I used Microsoft Vott to label the images and generate the dataset. 
Also, I am using Yolox3 tiny. 
How did I get rid of the NAN loss values when I am training with only 1 class?

**Comment by Robin2091** [comment]
Also, I just want to mention that my dataset has overlapping bounding boxes. Although, I am not sure if that affects anything.

**Comment by Robin2091** [comment]
@justOleh I got the training to work with 1 class. I added another tag to the dataset in Microsoft Vott (although our dataset had no instance of that tag) and I changed the num_classes flag to be 1. I also changed the class loss to a binary cross-entropy loss function and now the loss value does not give NANs.

**Comment by Yeah-Kun** [comment]
> @justOleh I got the training to work with 1 class. I added another tag to the dataset in Microsoft Vott (although our dataset had no instance of that tag) and I changed the num_classes flag to be 1. I also changed the class loss to a binary cross-entropy loss function and now the loss value does not give NANs.

hello, I have the same question with you, Can you tell me how to modify the loss function?

**Comment by Robin2091** [comment]
@Yeah-Kun Hi, so on on the models.py file, you will see this line under the YoloLoss function it is the last line before comment #6. 

class_loss = obj_mask * sparse_categorical_crossentropy(
            true_class_idx, pred_class)
Instead of sparse_categorical_crossentropy use binary_crossentropy(...). 

Also, change the num_classes FLAG on the train.py to 1 class. 

Then you have to re-export your dataset on Microsoft Vott with 2 tags. You don't have to have any instance of that tag(class) on your dataset. For some reason, when I export with 1 tag on Microsoft Vott it gives me NAN loss. But when I export it with 2 tags (even when I don't have any instance of that tag), I don't get NAN. So maybe try exporting with 1 tag. If it doesn't work then try with 2. This is the process that worked for me.

**Comment by Yeah-Kun** [comment]
> @ Yeah-Kun嗨，在models.py文件上，您将在YoloLoss函数下看到此行，它是注释[＃6](https://github.com/zzh8829/yolov3-tf2/issues/6)的最后一行。
> 
> class_loss = obj_mask * sparse_categorical_crossentropy（
> true_class_idx，pred_class）
> 而不是sparse_categorical_crossentropy，请使用binary_crossentropy（...）。
> 
> 另外，将train.py上的num_classes标志更改为1类。
> 
> 然后，您必须使用2个标签在Microsoft Vott上重新导出数据集。您不必在数据集中具有该标签（类）的任何实例。出于某种原因，当我在Microsoft Vott上使用1个标签导出时，会给我带来NAN损失。但是，当我用2个标签导出它时（即使我没有该标签的任何实例），也不会得到NAN。因此，也许尝试使用1个标签进行导出。如果不起作用，请尝试2。这是对我有用的过程。

Thanks your reply, I do the same as you suggest.
First, change the loss function
```
#class_loss = obj_mask * sparse_categorical_crossentropy(
#            true_class_idx, pred_class)

class_loss = obj_mask * binary_crossentropy(
            true_class_idx, pred_class)
```

Second, I modify the file `my_label.names`, add a new label (although I don't have any instance of that tag)
```
edge
nobody
```

but I still get the `nan` value in YOLOv3-Tiny and YOLOv3
```
      1/Unknown - 13s 13s/step - loss: nan - yolo_output_0_loss: nan - yolo_output_1_loss: 1441.6558
      2/Unknown - 18s 9s/step - loss: nan - yolo_output_0_loss: nan - yolo_output_1_loss: nan       
      3/Unknown - 23s 8s/step - loss: nan - yolo_output_0_loss: nan - yolo_output_1_loss: nan
      4/Unknown - 28s 7s/step - loss: nan - yolo_output_0_loss: nan - yolo_output_1_loss: nan
      5/Unknown - 33s 7s/step - loss: nan - yolo_output_0_loss: nan - yolo_output_1_loss: nan
      6/Unknown - 38s 6s/step - loss: nan - yolo_output_0_loss: nan - yolo_output_1_loss: nan
      7/Unknown - 44s 6s/step - loss: nan - yolo_output_0_loss: nan - yolo_output_1_loss: nan
      8/Unknown - 48s 6s/step - loss: nan - yolo_output_0_loss: nan - yolo_output_1_loss: nan
```

**Comment by Robin2091** [comment]
Hi, did you create your dataset on Microsoft Vott? You have to add the extra tag on Microsoft Vott and export your dataset with that extra tag. You do not have to change the names file.

**Comment by PieroCV** [comment]
> Hi, did you create your dataset on Microsoft Vott? You have to add the extra tag on Microsoft Vott and export your dataset with that extra tag. You do not have to change the names file.

Uhmmm... From vott, Did you export your annotations to Pascal Voc and then create a single tfrecord? Or just export to a group of tfrecords and then create the dataset?
Also, Did your model work?

**Comment by PieroCV** [comment]
Well, i managed to fix this issue!.
Is kind of cruel, but it works.
For dataset and val_dataset, i got some tfrecords 'corrupted', that make my model converge to NaN. These are Tfrecords imported from VoTT, so I'm looking for issues on VoTT, but everything seems OK there.
Test the train.py with `--mode fit --transfer fine_tune --classes PATH_CLASS` only, with 80 num of classes and same tfrecord for dataset and val_dataset. Try only two epochs for each .tfrecord in order to save time. You'll see a normal loss for some files and NaN loss for corrupted tfrecords.
The models.py is the original repo file. (sparse_categorical_crossentropy).
Tested and working.

**Comment by Robin2091** [comment]
@PieroCV  Hi, sorry for the late reply. I exported the dataset as individual tf records. I don't believe in my case the tf records were corrupted, because the training worked when i exported with 2 tags even when I didn't have any instance of the extra tag.

**Comment by zzh8829** [comment]
Looks like you have solved the issue. Please refer to this document https://github.com/zzh8829/yolov3-tf2/blob/master/docs/training_voc.md on transfer learning. I also introduce a dataset visualization tool to help debugging custom dataset issue.
